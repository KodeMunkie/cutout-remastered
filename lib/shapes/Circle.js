/**
 * A circle
 * @param {number} xBound maximum value for x-coordinates, zero-based
 * @param {number} yBound maximum value for y-coordinates, zero-based
 */
import fd from 'fdrandom';
import { toScanlines } from '../rasterizers/ellipse.js';
import { clampToInt, randomIntInclusive } from '../util.js';

export class Circle {
  constructor(xBound, yBound) {
    this.xBound = xBound;
    this.yBound = yBound;

    this.cx = randomIntInclusive(0, xBound);
    this.cy = randomIntInclusive(0, yBound);
    this.r = randomIntInclusive(1, 30);
  }

  set props([cx, cy, r]) {
    this.cx = cx;
    this.cy = cy;
    this.r = r;
  }

  get props() {
    return [this.cx, this.cy, this.r];
  }

  get svg() {
    const [cx, cy, r] = this.props;
    const shape = [
      'circle',
      {
        cx,
        cy,
        r
      }
    ];

    return shape;
  }

  clone() {
    const circle = new Circle(this.xBound, this.yBound);
    circle.props = this.props;

    return circle;
  }

  mutate() {
    /* istanbul ignore next */
    switch (randomIntInclusive(0, 1)) {
      case 0:
        this.cx = clampToInt(this.cx + fd.vrange(0,1,0.5) * 15, 0, this.xBound);
        this.cy = clampToInt(this.cy + fd.vrange(0,1,0.5) * 15, 0, this.yBound);
        break;
      case 1:
        this.r = clampToInt(this.r + fd.vrange(0,1,0.5) * 15, 1, this.xBound);
    }
  }

  rasterize() {
    const [cx, cy, rx] = this.props;
    const ry = rx;
    const angle = 0;

    return toScanlines(cx, cy, rx, ry, angle, this.xBound, this.yBound);
  }
}
