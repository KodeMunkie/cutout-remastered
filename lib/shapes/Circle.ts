import { toScanlines } from '../rasterizers/ellipse';
import { clampToInt, randomIntInclusive } from '../core/util';
import { Shape } from './Shape';
import { ShapeNameProps } from './ShapeNameProps';

export class Circle extends Shape {

  private cx: number;
  private cy: number;
  private r: number;

  constructor(xBound:number, yBound:number) {
    super(xBound, yBound);
    this.cx = randomIntInclusive(0, xBound);
    this.cy = randomIntInclusive(0, yBound);
    this.r = randomIntInclusive(1, 30);
  }

  set props([cx, cy, r]) {
    this.cx = cx;
    this.cy = cy;
    this.r = r;
  }

  get props() {
    return [this.cx, this.cy, this.r];
  }

  get svg(): ShapeNameProps {
    return {
      name: 'circle',
      props:
      {
        cx: this.cx,
        cy: this.cy,
        r: this.r
      }
    };
  }

  clone(): Shape {
    const circle: Circle = new Circle(this.xBound, this.yBound);
    circle.props = this.props;
    return circle;
  }

  mutate(): void {
    switch (randomIntInclusive(0, 1)) {
      case 0:
        this.cx = clampToInt(this.cx + this.random(), 0, this.xBound);
        this.cy = clampToInt(this.cy + this.random(), 0, this.yBound);
        break;
      case 1:
        this.r = clampToInt(this.r + this.random(), 1, this.xBound);
    }
  }

  rasterize(): number[][] {
    return toScanlines(this.cx, this.cy, this.r, this.r, 0, this.xBound, this.yBound);
  }
}
