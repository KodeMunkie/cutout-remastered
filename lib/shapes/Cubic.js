const { randomIntInclusive, clampToInt, randomNormal } = require('../util');
const { bezierToScanlines } = require('../rasterizer');

/**
 * A cubic bezier
 * @param {number} xBound maximum value for x-coordinates, zero-based
 * @param {number} yBound maximum value for y-coordinates, zero-based
 */

class Cubic {
  constructor(xBound, yBound) {
    this.xBound = xBound;
    this.yBound = yBound;

    this.x1 = randomIntInclusive(0, xBound);
    this.y1 = randomIntInclusive(0, yBound);
    this.x2 = this.x1 + randomIntInclusive(-15, 15);
    this.y2 = this.y1 + randomIntInclusive(-15, 15);
    this.x3 = this.x1 + randomIntInclusive(-15, 15);
    this.y3 = this.y1 + randomIntInclusive(-15, 15);
    this.x4 = this.x1 + randomIntInclusive(-15, 15);
    this.y4 = this.y1 + randomIntInclusive(-15, 15);
  }

  set props([x1, y1, x2, y2, x3, y3, x4, y4]) {
    this.x1 = x1;
    this.y1 = y1;
    this.x2 = x2;
    this.y2 = y2;
    this.x3 = x3;
    this.y3 = y3;
    this.x4 = x4;
    this.y4 = y4;
  }

  get props() {
    return [this.x1, this.y1, this.x2, this.y2, this.x3, this.y3, this.x4, this.y4];
  }

  get svg() {
    const [x1, y1, x2, y2, x3, y3, x4, y4] = this.props;
    const shape = [
      'path',
      {
        d: `M${x1},${y1} C${x2},${y2} ${x3},${y3} ${x4},${y4}`
      }
    ];

    return shape;
  }

  clone() {
    const cubic = new Cubic(this.xBound, this.yBound);
    cubic.props = this.props;

    return cubic;
  }

  mutate() {
    /* istanbul ignore next */
    switch (randomIntInclusive(0, 3)) {
      case 0:
        this.x1 = clampToInt(this.x1 + randomNormal() * 15, 0, this.xBound);
        this.y1 = clampToInt(this.y1 + randomNormal() * 15, 0, this.yBound);
        break;
      case 1:
        this.x2 = clampToInt(this.x2 + randomNormal() * 15, 0, this.xBound);
        this.y2 = clampToInt(this.y2 + randomNormal() * 15, 0, this.yBound);
        break;
      case 2:
        this.x3 = clampToInt(this.x3 + randomNormal() * 15, 0, this.xBound);
        this.y3 = clampToInt(this.y3 + randomNormal() * 15, 0, this.yBound);
        break;
      case 3:
        this.x4 = clampToInt(this.x4 + randomNormal() * 15, 0, this.xBound);
        this.y4 = clampToInt(this.y4 + randomNormal() * 15, 0, this.yBound);
    }
  }

  rasterize() {
    return bezierToScanlines(this.props, this.xBound, this.yBound);
  }
}

module.exports = Cubic;
