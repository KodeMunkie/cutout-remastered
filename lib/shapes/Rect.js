/**
 * A rectangle
 * @param {number} xBound maximum value for x-coordinates, zero-based
 * @param {number} yBound maximum value for y-coordinates, zero-based
 */
import { clampToInt, randomIntInclusive } from '../util.js';
import fd from 'fdrandom';
import { toScanlines } from '../rasterizers/rectangle.js';

export class Rect {
  constructor(xBound, yBound) {
    this.xBound = xBound;
    this.yBound = yBound;

    this.x = randomIntInclusive(0, xBound);
    this.y = randomIntInclusive(0, yBound);
    this.width = clampToInt(randomIntInclusive(1, 30), 1, xBound - this.x);
    this.height = clampToInt(randomIntInclusive(1, 30), 1, yBound - this.y);
  }

  set props([x, y, width, height]) {
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
  }

  get props() {
    return [this.x, this.y, this.width, this.height];
  }

  get svg() {
    const [x, y, width, height] = this.props;
    const shape = [
      'rect',
      {
        x,
        y,
        width,
        height
      }
    ];

    return shape;
  }

  clone() {
    const rectangle = new Rect(this.xBound, this.yBound);
    rectangle.props = this.props;

    return rectangle;
  }

  mutate() {
    /* istanbul ignore next */
    switch (randomIntInclusive(0, 1)) {
      case 0:
        this.x = clampToInt(this.x + fd.vrange(0,1,0.5) * 15, 0, this.xBound);
        this.y = clampToInt(this.y + fd.vrange(0,1,0.5) * 15, 0, this.yBound);
        break;
      case 1:
        this.width = clampToInt(this.width + fd.vrange(0,1,0.5) * 15, 1, this.xBound - this.x);
        this.height = clampToInt(this.height + fd.vrange(0,1,0.5) * 15, 1, this.yBound - this.y);
    }
  }

  rasterize() {
    const [x, y, width, height] = this.props;

    return toScanlines(x, y, width, height, this.xBound, this.yBound);
  }
}
