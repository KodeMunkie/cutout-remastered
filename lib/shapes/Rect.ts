/**
 * A rectangle
 * @param {number} xBound maximum value for x-coordinates, zero-based
 * @param {number} yBound maximum value for y-coordinates, zero-based
 */
//@ts-ignore
import fd from 'fdrandom';
import { clampToInt, randomIntInclusive } from '../core/util';
import { toScanlines } from '../rasterizers/rectangle';
import { Shape } from './Shape';

export class Rect extends Shape {

  private x: number;
  private y: number;
  private width: number;
  private height: number;

  constructor(xBound: number, yBound: number) {
    super(xBound, yBound);
    this.x = randomIntInclusive(0, xBound);
    this.y = randomIntInclusive(0, yBound);
    this.width = clampToInt(randomIntInclusive(1, 30), 1, xBound - this.x);
    this.height = clampToInt(randomIntInclusive(1, 30), 1, yBound - this.y);
  }

  set props([x, y, width, height]) {
    this.x = x;
    this.y = y;
    this.width = width;
    this.height = height;
  }

  get props() {
    return [this.x, this.y, this.width, this.height];
  }

  get svg() {
    const [x, y, width, height] = this.props;
    const shape = [
      'rect',
      {
        x,
        y,
        width,
        height
      }
    ];

    return shape;
  }

  clone() {
    const rectangle = new Rect(this.xBound, this.yBound);
    rectangle.props = this.props;

    return rectangle;
  }

  mutate() {
    /* istanbul ignore next */
    switch (randomIntInclusive(0, 1)) {
      case 0:
        this.x = clampToInt(this.x + fd.vrange(0,1,0.5) * 15, 0, this.xBound);
        this.y = clampToInt(this.y + fd.vrange(0,1,0.5) * 15, 0, this.yBound);
        break;
      case 1:
        this.width = clampToInt(this.width + fd.vrange(0,1,0.5) * 15, 1, this.xBound - this.x);
        this.height = clampToInt(this.height + fd.vrange(0,1,0.5) * 15, 1, this.yBound - this.y);
    }
  }

  rasterize() {
    const [x, y, width, height] = this.props;

    return toScanlines(x, y, width, height, this.xBound, this.yBound);
  }
}