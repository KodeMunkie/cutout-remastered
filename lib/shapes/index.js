const debug = require('debug')('cutout:shapes');
const { randomArrayItem } = require('../util');
const Rect = require('./Rect');

/**
 * Create a shape of the supplied type
 */

const create = (shapeType, xBound, yBound) => {
  debug(`Creating ${shapeType}`);

  switch (shapeType.toLowerCase()) {
    case 'rect':
      return new Rect(xBound, yBound);
    default:
      throw new Error(`${shapeType} is not a valid shape type`);
  }
};

/**
 * Creates a random shape from the types supplied
 */

const randomShapeOf = (shapeTypes, xBound, yBound) =>
  create(randomArrayItem(shapeTypes), xBound, yBound);

module.exports = { randomShapeOf };
