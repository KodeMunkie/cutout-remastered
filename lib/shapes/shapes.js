import { Circle } from './Circle.js';
import { RotatedEllipse }from './RotatedEllipse.js';
import { Ellipse } from './Ellipse.js';
import { Square } from './Square.js';
import { Rect } from './Rect.js';
import { RotatedRect } from './RotatedRect.js';
import { Triangle } from './Triangle.js';
import { Line } from './Line.js';
import { Quadratic } from './Quadratic.js';
import { Cubic } from './Cubic.js';
import { randomArrayItem } from '../util.js';

/**
 * Create a shape of the supplied type
 */

export const create = (shapeType, xBound, yBound) => {
  switch (shapeType.toLowerCase()) {
    case 'circle':
      return new Circle(xBound, yBound);
    case 'rotatedellipse':
      return new RotatedEllipse(xBound, yBound);
    case 'ellipse':
      return new Ellipse(xBound, yBound);
    case 'square':
      return new Square(xBound, yBound);
    case 'rect':
      return new Rect(xBound, yBound);
    case 'rotatedrect':
      return new RotatedRect(xBound, yBound);
    case 'triangle':
      return new Triangle(xBound, yBound);
    case 'line':
      return new Line(xBound, yBound);
    case 'quadratic':
      return new Quadratic(xBound, yBound);
    case 'cubic':
      return new Cubic(xBound, yBound);
    default:
      throw new Error(`${shapeType} is not a valid shape type`);
  }
};

/**
 * Creates a random shape from the types supplied
 */

export const randomShapeOf = (shapeTypes, xBound, yBound) =>
  create(randomArrayItem(shapeTypes), xBound, yBound);
