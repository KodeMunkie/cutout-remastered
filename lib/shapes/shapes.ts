import { Circle } from './Circle';
import { RotatedEllipse }from './RotatedEllipse';
import { Ellipse } from './Ellipse';
import { Square } from './Square';
import { Rect } from './Rect';
import { RotatedRect } from './RotatedRect';
import { Triangle } from './Triangle';
import { Line } from './Line';
import { Quadratic } from './Quadratic';
import { Cubic } from './Cubic';
import { randomArrayItem } from '../core/util';
import { Shape } from './Shape';

/**
 * Create a shape of the supplied type
 */
export const create = (shapeType: string, xBound: number, yBound: number): Shape => {
  switch (shapeType.toLowerCase()) {
    case 'circle':
      return new Circle(xBound, yBound);
    case 'rotatedellipse':
      return new RotatedEllipse(xBound, yBound);
    case 'ellipse':
      return new Ellipse(xBound, yBound);
    case 'square':
      return new Square(xBound, yBound);
    case 'rect':
      return new Rect(xBound, yBound);
    case 'rotatedrect':
      return new RotatedRect(xBound, yBound);
    case 'triangle':
      return new Triangle(xBound, yBound);
    case 'line':
      return new Line(xBound, yBound);
    case 'quadratic':
      return new Quadratic(xBound, yBound);
    case 'cubic':
      return new Cubic(xBound, yBound);
    default:
      throw new Error(`${shapeType} is not a valid shape type`);
  }
};

/**
 * Creates a random shape from the types supplied
 */

export const randomShapeOf = (shapeTypes: string[], xBound: number, yBound: number): Shape =>
  create(randomArrayItem(shapeTypes), xBound, yBound);
