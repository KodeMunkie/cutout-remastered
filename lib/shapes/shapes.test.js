import { Rect } from './Rect.js';
import { randomShapeOf } from './shapes.js';

it('returns a shape', () => {
  const xBound = 10;
  const yBound = 10;
  const shape = randomShapeOf(['Rect'], xBound, yBound);

  expect(shape).toBeInstanceOf(Rect);
});

it('throws an error for invalid shape types', () => {
  const xBound = 10;
  const yBound = 10;

  expect(() => randomShapeOf(['Rhombus'], xBound, yBound)).toThrowErrorMatchingSnapshot();
});
