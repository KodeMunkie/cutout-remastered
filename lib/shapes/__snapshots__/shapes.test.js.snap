// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws an error for invalid shape types 1`] = `"Rhombus is not a valid shape type"`;
